src/api/app/models/staging/stage_requests.rb

```diff
@@ -0,0 +1,38 @@
+class Staging::StageRequests
+  include ActiveModel::Model
+  attr_accessor :requests, :staging_project_name
+
+  def perform
+    self.result = []
+    requests.each do |request|
+      bs_request_action = request.bs_request_actions.first
+      if bs_request_action.is_submit?
+        branch_package(bs_request_action)
+      elsif bs_request_action.is_delete?
+        # TODO: implement delete requests
+      end
+      result
+    end
+  end
+
+  private
+
+  attr_accessor :result
+
+  def branch_package(bs_request_action)
+    request = bs_request_action.bs_request
+    BranchPackage.new(
+      target_project: staging_project_name,
+      target_package: bs_request_action.target_package,
+      project: bs_request_action.source_project,
+      package: bs_request_action.source_package,
+      extend_package_names: false
+    ).branch
+    result << request
+  rescue BranchPackage::DoubleBranchPackageError
+    # we leave the package and just assign the request
+    result << request
```

Comment:
The difference between the two exceptions is that the BranchPackage::DoubleBranchPackageError is expected, given that some user already created that package in the target project.
So we don't need to tell the OBS developers. But we still should let the user know about this. It could be that the content of the two packages differ.

tldr;

I would collect successfully branched packages and failures separately and let users know which packages failed. Ideally with a failure message. Right now that would be $package_already_exist or $internal_server_error.


Probably out of scope for now now, but we probably want to make this "branching report" stick (until a user dismisses it). Right now, users that close their browser after sending the request won't have a chance to see the error message. Same goes for other users that maintain the project, but didn't send the request.

Reply:
Fixed